### ISSUE TITLE
Add and replace / Add, replace and play does not work

### ISSUE THREAD

**Issue by kristianbenoit** [abarisain/dmix#701]
Since I upgraded to 1.07, I can no longer "long click on item in library / Add and replace" and "longclick on item in library / Add, replace and play". It just do nothing, the music keeps playing, nothing is added to the current playlist and no messages pop at the bottom of the screen.

Server version is reported as 0.15.0 by `mpdroid` and 0.15.15 by `mpd -V` it runs on TomatoUSB and is part of optware (http://www.nslu2-linux.org/wiki/Optware/HomePage) The package can be found here http://ipkg.nslu2-linux.org/feeds/optware/ddwrt/cross/stable/mpd_0.15.15-2_mipsel.ipk. These binary are for mips. This seems to be the source (http://sources.nslu2-linux.org/sources/mpd-0.15.15.tar.bz2).

**Comment by avuton** [comment]
Ugh, I thought I checked this a week or so ago, I'll try it again soon. If I can't find the problem, I'm going to need ```logcat``` output.

**Comment by kristianbenoit** [comment]
I thought it was so obvious that it would be fix quickly, but it might be
with my specific mpd version.

Which tags should I look for in logcat ?

**Comment by avuton** [comment]
Well, I checked all versions back to 0.15.x a week ago (for an unrelated bug) and I actually used clear, add, replace to check them all (though I might have gotten a little lax by the point I got to 0.15.x).

Well, you should definitely be getting an exception of some sort, that exception would be very interesting.

**Comment by kristianbenoit** [comment]
Here's the log:

E/MPDConnectionMultiSocket( 5250): Command command_list_begin
E/MPDConnectionMultiSocket( 5250): moveid "1004" "0"
E/MPDConnectionMultiSocket( 5250): delete "1:37"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/01. Combustion lente.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/02. Se perdre.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/03. Le Cœur dans la tête.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/04. Montréal.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/05. Retourne chez elle.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/06. Farine 5 roses.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/07. Imparfait.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/08. Will You Follow Me.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/09. Terminus.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/10. Laboratoire amoureux.ogg"
E/MPDConnectionMultiSocket( 5250): add "Ariane Moffatt/Le Cœur dans la tête/11. Histoire d'ère.ogg"
E/MPDConnectionMultiSocket( 5250): next
E/MPDConnectionMultiSocket( 5250): delete "0"
E/MPDConnectionMultiSocket( 5250): command_list_end failed after 4 attempts.
E/MPDConnectionMultiSocket( 5250): org.a0z.mpd.exception.MPDException: ACK [2@1] {delete} need a positive integer
E/MPDConnectionMultiSocket( 5250): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.read(MPDConnection.java:602)
E/MPDConnectionMultiSocket( 5250): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:396)
E/MPDConnectionMultiSocket( 5250): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:364)
E/MPDConnectionMultiSocket( 5250): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/MPDConnectionMultiSocket( 5250): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
E/MPDConnectionMultiSocket( 5250): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
E/MPDConnectionMultiSocket( 5250): 	at java.lang.Thread.run(Thread.java:841)
E/ArtistsFragment( 5250): Failed to add to queue.
E/ArtistsFragment( 5250): org.a0z.mpd.exception.MPDException: ACK [2@1] {delete} need a positive integer
E/ArtistsFragment( 5250): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.read(MPDConnection.java:602)
E/ArtistsFragment( 5250): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:396)
E/ArtistsFragment( 5250): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:364)
E/ArtistsFragment( 5250): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/ArtistsFragment( 5250): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
E/ArtistsFragment( 5250): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
E/ArtistsFragment( 5250): 	at java.lang.Thread.run(Thread.java:841)

**Comment by avuton** [comment]
Owh(!) Looks like 0.15.x doesn't support ranges. Very very useful, thanks. I hope to have a test build soon for you.

**Comment by avuton** [comment]
My suspicion was correct:

```
commit 31cabc751d2dc6721706d878a0694afc83ef6a0c
Author: Max Kellermann <max@duempel.org>
Date:   Wed Sep 30 23:13:13 2009 +0200

    command: range support for "delete"
```

[Here's a test build](https://drive.google.com/file/d/0B1aD76XF-iKLbUVxamRrYmxiWUU/view?usp=sharing). This is a backport to that version, this will be different (and fixed) in the next version as well. The caveat is this will ultimately fail on large playlist queues (150+ items), and will be broken until I get around to implementing split command queues. So; if it fails on large playlists, do not report it; once I implement split command queues, it will not be able to be backported. Also, once you change to this build you will not be automatically updated to the next version of MPDroid, so, keep this in mind.

**Comment by avuton** [comment]
[Updated build](https://drive.google.com/file/d/0B1aD76XF-iKLdm1IOVlzdmNpcGM/view?usp=sharing); sorry the first build was broken, this one should actually work.

**Comment by abarisain** [comment]
A little offtopic but : Should we still support such old MPD versions ?

**Comment by avuton** [comment]
Absolutely, it's not mandatory by any means, but backwards compatibility is (typically) trivial.

**Comment by kristianbenoit** [comment]
@abarisain, this is the only version provided by optware which is a package distribution for routers. The nice thing about MPD is to be able to use it in weird devices such as a router. Having the best interface to manipulate it essential.

Still not working:
```
E/MPDConnectionMultiSocket( 4536): Command command_list_begin
E/MPDConnectionMultiSocket( 4536): moveid "1023" "0"
E/MPDConnectionMultiSocket( 4536): delete "37"
E/MPDConnectionMultiSocket( 4536): delete "36"
E/MPDConnectionMultiSocket( 4536): delete "35"
E/MPDConnectionMultiSocket( 4536): delete "34"
E/MPDConnectionMultiSocket( 4536): delete "33"
E/MPDConnectionMultiSocket( 4536): delete "32"
E/MPDConnectionMultiSocket( 4536): delete "31"
E/MPDConnectionMultiSocket( 4536): delete "30"
E/MPDConnectionMultiSocket( 4536): delete "29"
E/MPDConnectionMultiSocket( 4536): delete "28"
E/MPDConnectionMultiSocket( 4536): delete "27"
E/MPDConnectionMultiSocket( 4536): delete "26"
E/MPDConnectionMultiSocket( 4536): delete "25"
E/MPDConnectionMultiSocket( 4536): delete "24"
E/MPDConnectionMultiSocket( 4536): delete "23"
E/MPDConnectionMultiSocket( 4536): delete "22"
E/MPDConnectionMultiSocket( 4536): delete "21"
E/MPDConnectionMultiSocket( 4536): delete "20"
E/MPDConnectionMultiSocket( 4536): delete "19"
E/MPDConnectionMultiSocket( 4536): delete "18"
E/MPDConnectionMultiSocket( 4536): delete "17"
E/MPDConnectionMultiSocket( 4536): delete "16"
E/MPDConnectionMultiSocket( 4536): delete "15"
E/MPDConnectionMultiSocket( 4536): delete "14"
E/MPDConnectionMultiSocket( 4536): delete "13"
E/MPDConnectionMultiSocket( 4536): delete "12"
E/MPDConnectionMultiSocket( 4536): delete "11"
E/MPDConnectionMultiSocket( 4536): delete "10"
E/MPDConnectionMultiSocket( 4536): delete "9"
E/MPDConnectionMultiSocket( 4536): delete "8"
E/MPDConnectionMultiSocket( 4536): delete "7"
E/MPDConnectionMultiSocket( 4536): delete "6"
E/MPDConnectionMultiSocket( 4536): delete "5"
E/MPDConnectionMultiSocket( 4536): delete "4"
E/MPDConnectionMultiSocket( 4536): delete "3"
E/MPDConnectionMultiSocket( 4536): delete "2"
E/MPDConnectionMultiSocket( 4536): delete "1"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/01. Brandon Marlow.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/01 Brandon Marlow.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/02. Gumshoe (Featuring Beni Bbq).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/02 Gumshoe (Featuring Beni Bbq).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/03. Ave Mucho (Featuring Bran Van 3000).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/03 Ave Mucho (Featuring Bran Van 3000).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/04. November Number 3 (Featuring Fanny Bloom).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/04 November Number 3 (Featuring Fanny Bloom).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/05. Dan Dan.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/05 Dan Dan.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/06. Mojo Ego (Featuring Gigi French).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/06 Mojo Ego (Featuring Gigi French).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/07. Lillehammer.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/07 Lillehammer.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/08. Mama Donte (Featuring Senja Sargeant).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/08 Mama Donte (Featuring Senja Sargeant).mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/09. Monster Donte.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/09 Monster Donte.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/10. Sweet Charlemagne.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/10 Sweet Charlemagne.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire (MV)/Golden Bombay/11. Gordon Bombay.mp3"
E/MPDConnectionMultiSocket( 4536): add "Misteur Valaire/Misteur Valaire/Golden Bombay/11 Gordon Bombay.mp3"
E/MPDConnectionMultiSocket( 4536): next
E/MPDConnectionMultiSocket( 4536): delete "0"
E/MPDConnectionMultiSocket( 4536): command_list_end failed after 4 attempts.
E/MPDConnectionMultiSocket( 4536): org.a0z.mpd.exception.MPDException: ACK [2@1] {delete} Bad song index
E/MPDConnectionMultiSocket( 4536): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.read(MPDConnection.java:602)
E/MPDConnectionMultiSocket( 4536): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:396)
E/MPDConnectionMultiSocket( 4536): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:364)
E/MPDConnectionMultiSocket( 4536): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/MPDConnectionMultiSocket( 4536): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
E/MPDConnectionMultiSocket( 4536): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
E/MPDConnectionMultiSocket( 4536): 	at java.lang.Thread.run(Thread.java:841)
E/AlbumsFragment( 4536): Failed to add.
E/AlbumsFragment( 4536): org.a0z.mpd.exception.MPDException: ACK [2@1] {delete} Bad song index
E/AlbumsFragment( 4536): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.read(MPDConnection.java:602)
E/AlbumsFragment( 4536): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:396)
E/AlbumsFragment( 4536): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:364)
E/AlbumsFragment( 4536): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/AlbumsFragment( 4536): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
E/AlbumsFragment( 4536): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
E/AlbumsFragment( 4536): 	at java.lang.Thread.run(Thread.java:841)
```

**Comment by avuton** [comment]
Whoops, off by one error. Guess I should have tested it :) new test build when I get home.

**Comment by avuton** [comment]
Also: That's exactly why I wouldn't want to drop support for older versions of the protocol. People can't always help how old their backend is, but they should be able to control the frontend, at least. Anyhow [here's a newer test build](https://drive.google.com/file/d/0B1aD76XF-iKLWUcyVmExcksydTQ/view?usp=sharing), let me know if it's still broken, but it definitely works here with 0.15.16.

**Comment by kristianbenoit** [comment]
It works fine when the music is stopped, but still have an error when the music is playing...

```
E/MPDConnectionMultiSocket(14958): Command command_list_begin
E/MPDConnectionMultiSocket(14958): moveid "1041" "0"
E/MPDConnectionMultiSocket(14958): delete "1:14"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/A_Voltado_Malandro.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/Mon_Amour.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/Photo.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/Barbara.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/Estate.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/La_Nuit_Des_Masques.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/La_Gran_Périda_De_Alhama.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/Le_Miroire_Aux_Oiseaux.ogg"
E/MPDConnectionMultiSocket(14958): add "Bia/Lammoireduvent/La_Tour_De_Constance.ogg"
E/MPDConnectionMultiSocket(14958): add "Various Artists/Putumayo Presents_ Brazilian Lounge/06. Mariana.ogg"
E/MPDConnectionMultiSocket(14958): next
E/MPDConnectionMultiSocket(14958): delete "0"
E/MPDConnectionMultiSocket(14958): command_list_end failed after 4 attempts.
E/MPDConnectionMultiSocket(14958): org.a0z.mpd.exception.MPDException: ACK [2@1] {delete} need a positive integer
E/MPDConnectionMultiSocket(14958): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.read(MPDConnection.java:602)
E/MPDConnectionMultiSocket(14958): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:396)
E/MPDConnectionMultiSocket(14958): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:364)
E/MPDConnectionMultiSocket(14958): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/MPDConnectionMultiSocket(14958): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
E/MPDConnectionMultiSocket(14958): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
E/MPDConnectionMultiSocket(14958): 	at java.lang.Thread.run(Thread.java:841)
E/ArtistsFragment(14958): Failed to add to queue.
E/ArtistsFragment(14958): org.a0z.mpd.exception.MPDException: ACK [2@1] {delete} need a positive integer
E/ArtistsFragment(14958): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.read(MPDConnection.java:602)
E/ArtistsFragment(14958): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:396)
E/ArtistsFragment(14958): 	at org.a0z.mpd.connection.MPDConnection$CommandProcessor.call(MPDConnection.java:364)
E/ArtistsFragment(14958): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/ArtistsFragment(14958): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
E/ArtistsFragment(14958): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
E/ArtistsFragment(14958): 	at java.lang.Thread.run(Thread.java:841)
```

Thank you !

**Comment by avuton** [comment]
[One more try?](https://drive.google.com/file/d/0B1aD76XF-iKLbDZCTmVWTGlESEU/view?usp=sharing) I don't know what happened, but I have to assume it was a upload from a wrong place or something; it works perfectly here on 0.15.16. Hopefully this time I got it right :+1:

**Comment by kristianbenoit** [comment]
Wow, nice job! Works perfectly.

Thanks ! And for all the other that wont notice.

**Comment by avuton** [comment]
Fixed via 3a26fc8.
